soco/events_base.py

```diff
@@ -0,0 +1,642 @@
+                    self.service.base_url +
+                    self.service.event_subscription_url,
+                    self.sid))
+
+        self.request(
+            'SUBSCRIBE',
+            self.service.base_url + self.service.event_subscription_url,
+            headers, success, failure)
+
+    def unsubscribe(self):
+        """unsubscribe()
+        Unsubscribe from the service's events.
+        Once unsubscribed, a Subscription instance should not be reused
+        """
+        # Trying to unsubscribe if already unsubscribed, or not yet
+        # subscribed, fails silently
+        if self._has_been_unsubscribed or not self.is_subscribed:
+            return
+        self._cancel_subscription()
+
+        # If the subscription has timed out, an attempt to
+        # unsubscribe from it will fail.
+        if self.time_left == 0:
+            log.info(
+                "Subscription %s, sid: %s has expired. No need to unsubscribe",
+                self.service.base_url + self.service.event_subscription_url,
+                self.sid)
+            return
+
+        # Send an unsubscribe request like this:
+        # UNSUBSCRIBE publisher path HTTP/1.1
+        # HOST: publisher host:publisher port
+        # SID: uuid:subscription UUID
+        headers = {
+            'SID': self.sid
+        }
+
+        # pylint: disable=missing-docstring, unused-argument
+        def success(*arg):
+            log.info(
+                "Unsubscribed from %s, sid: %s",
+                self.service.base_url + self.service.event_subscription_url,
+                self.sid)
+
+        # pylint: disable=missing-docstring
+        def failure():
+            log.warning(
+                "Error attempting to unsubscribe from %s, sid: %s",
+                self.service.base_url + self.service.event_subscription_url,
+                self.sid)
+
+        self.request(
+            'UNSUBSCRIBE',
+            self.service.base_url + self.service.event_subscription_url,
+            headers, success, failure)
+
+    def send_to_service(self, event):
+        """Send an `Event` to self.callback or self.events.
+        If self.callback is set and is callable, it will be called with the
+        `Event` as the only parameter. Otherwise the `Event` will be sent to
+        self.events. As self.callback is not threadsafe, it should be set
+        only if `soco.events_twisted.Subscription` is being used.
+
+        Args:
+            event(Event): The `Event` to send to self.callback or
+                self.events.
+
+        """
+        if hasattr(self, 'callback'):
+            # pylint: disable=no-member
+            callback = self.callback
+        else:
+            callback = None
+        if callback and hasattr(callback, '__call__'):
+            callback(event)
+        else:
+            self.events.put(event)
+
+    # pylint: disable=missing-docstring
+    def auto_renew_start(self, interval):
+        pass
```

Comment:
Couldn't this method be private (underscored)? It would help to know which methods get called from the outside and which are just subroutines.

Reply:
Will do.

Reply:
Actually, I was just referring to the abstract Subscription methods `request`, `auto_renew_start` and `auto_renew_cancel`. These only get called by other methods within the same class, so there should be no problem making them private.